### DISCUSSION TITLE
How to extend null-ls sources

### DISCUSSION THREAD

**Discussion by dpetka2001** [LazyVim/LazyVim#154] [Q&A] [answered]
I tried to extend null-ls sources with the following code
```lua
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      return vim.list_extend(opts.sources, { "require('null-ls').builtins.formatting.black" })
    end,
  },
```
but `black` doesn't seem to be working. I get this error `[LSP] Format request failed, no matching language servers`.
`black` is correctly installed via mason and if I use 
```lua
{
  "jose-elias-alvarez/null-ls.nvim",
  opts = function()
    local nls = require("null-ls")
    return {
      sources = {
        -- nls.builtins.formatting.prettierd,
        nls.builtins.formatting.black,
      },
    }
  end,
}
```
it correctly attaches `black` but `stylua` and `flake8` get lost from the configuration and then 
I cannot properly format `.lua` files.
What is the correct way of extending the null-ls sources?

**Comment by dpetka2001** [comment]
Ok so I solved it with 
```lua
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function()
      local nls = require("null-ls")
      return nls.register({ nls.builtins.formatting.black })
    end,
  },
```
I don't know if there is another way but it seems that it's working properly now without overriding the default
sources of `null-ls` set by LazyVim.

**Comment by dpetka2001** [comment, marked as answer]
After further examining the docs I found the proper(?) way to configure it according to the doc 
examples:
```lua
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, { nls.builtins.formatting.black })
    end,
  }
```
Thank you for the great documentation. Even as a beginner you still have to dig into Neovim Lua
to properly customize your own configuration and it's really fun and the LazyVim site gives great 
examples as to how to find your way.

**Reply by lsshawn** [reply]
I tried this but when i run `:checkhealth`, null-ls is saying no sources registered. Any experience?

```
  {
    "jose-elias-alvarez/null-ls.nvim",
    event = "BufReadPre",
    dependencies = { "mason.nvim" },
    opts = function()
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, {
        nls.builtins.formatting.prettierd.with({
          extra_args = {
            "--trailing-comma=all",
            "--single-quote=true",
            "--semi=true",
            "--tab-width=2",
            "--print-width=80",
            "--single-attribute-per-line=true",
            "--comma-dangle=always-multiline",
          },
        }),
        nls.builtins.formatting.stylua,
        nls.builtins.diagnostics.flake8,
      })
    end,
  },
```

**Reply by folke** [reply]
@lsshawn your code is wrong. You forgot the `opts` parameter to the `opts` function.
